### PR TITLE
Feat/alert/edit alert navigation

### PR DESCRIPTION
**Pull request by francelu**
# Implement Navigation to Edit Alert Screen

## Description
This PR introduces the navigation to `EditAlertScreen`. It closes issue #239 .  
I had asked help from @agonzalez-r with the problem of the update not being updated. She couldn't figure it out beside it was a VM / Supabase problem. In the end, I found out we had no RLS for updating in Supabase.

## Changes
`EditAlertScreen` has necessary changes implemented in relation to the navigation to it.  
UI ways, `EditAlertScreen` now has a reset button when it's running.

#### Navigation Way
- `EditAlertScreen` takes in only parameters `locationViewModel`, `gpsService`, `alertViewModel` and `navigationActions.
- `MainActivity` has the possibility to navigate to `EditAlertScreen` in the `Route.ALERT_LIST`.
- `NavigationActions` has a new `Screen.EDIT_ALERT`.
- `AlertListsScreen` has edit button `onClick` action changed (selects the alert to edit and navigates).

#### Alert VM to Handle Which Alert to Edit
- `AlertViewModel` has a public `State` `selectedEditAlert` which is set when `selectedEditAlert()` is called.
- `AlertViewModelTest.kt` has been updated accordingly.

#### Helpful
- Added `convertToProduct()` and `convertToUrgency()` functions in `AlertComponents.kt` to be able to update the product and urgency on value changes.

#### Added
- None.

#### Modified
- `app/src/main/java/com/android/periodpals/MainActivity.kt`
- `app/src/main/java/com/android/periodpals/model/alert/AlertViewModel.kt`
- `app/src/main/java/com/android/periodpals/ui/alert/AlertLists.kt`
- `app/src/main/java/com/android/periodpals/ui/alert/EditAlert.kt`
- `app/src/main/java/com/android/periodpals/components/AlertComponents.kt`
- `app/src/main/java/com/android/periodpals/navigation/NavigationActions.kt`
- `app/test/java/com/android/periodpals/model/alert/AlertViewModelTest.kt`
- `app/test/java/com/android/periodpals/ui/alert/AlertListsScreenTest.kt`
- `app/test/java/com/android/periodpals/ui/alert/EditAlertScreenTest.kt`
- `app/test/java/com/android/periodpals/ui/navigation/NavigationActionsTest.kt`

#### Removed
- None.

## Dependencies Added
- None.

## Testing
- `AlertViewModelTest.kt`: added a test for the `selectEditAlert()` function.
- `AlertListsScreenTest.kt`: added a test to verify the navigation to `EditAlertScreen`.
- `EditAlertScreenTest.kt`: updated the mocking and tests to include the verification of the calls to supabase and so forth.
- `NavigationActionsTest.kt`: moved it to the right place.

## Screenshots
- Not relevant.

### CODE REVIEW

File: app/src/test/java/com/android/periodpals/ui/alert/EditAlertScreenTest.kt [hunk 1]
```diff
@@ -62,30 +69,49 @@ class EditAlertScreenTest {
     private const val DELETE_BUTTON_TEXT = "Delete"
     private const val SAVE_BUTTON_TEXT = "Save"
     private const val RESOLVE_BUTTON_TEXT = "Resolve"
+
+    private const val NUM_ITEMS_WHEN_SUGGESTION = 4
```
Comment (agonzalez-r) [inline app/src/test/java/com/android/periodpals/ui/alert/EditAlertScreenTest.kt:73]: remove these if they are not used

**Review by agonzalez-r — CHANGES_REQUESTED** [review]
## Summary
Very good job! I know this task required solving some bugs with the repository, kudos for solving them so quickly. 
## Important 
#### Code Quality
- Overall, code is very clean.
- I left a comment about some renaming for better usability. 
#### Functionality
- The error handling for the `selectedEditAlert` being null in `EditAlert` is not appropriate and needs to be changed. 
- The functions you added in `AlertComponents.kt` are basically the same as the ones I had added (and tested) in `Alert.kt`, I suggest removing the ones you added, but we could also discuss keeping the ones you added and removing mine. TBD. Either way, let's not keep both versions. 
#### Performance
- I tested manually and everything works as expected. Great!!
## Testing
The tests you added are correct. Everything passes!
## Documentation
Very complete, too complete, you even completed missing documentation in other files! 
## Steps before PR approved
- Change error handling at the beginning of `EditAlertScreen()`
- Remove functions added to `AlertComponents` as they are not necessary

File: app/src/main/java/com/android/periodpals/model/alert/AlertViewModel.kt [hunk 2]
```diff
@@ -182,4 +187,13 @@ class AlertViewModel(private val alertModelSupabase: AlertModelSupabase) : ViewM
   fun setUserID(uid: String) {
     viewModelScope.launch { userId.value = uid }
   }
+
+  /**
+   * Selects an alert to be edited.
+   *
+   * @param alert The alert to be edited.
+   */
+  fun selectEditAlert(alert: Alert) {
```
[the remark this answers is not in the document]
Comment (francelu) [reply app/src/main/java/com/android/periodpals/model/alert/AlertViewModel.kt:196]: I did the refactorings :)

File: app/src/main/java/com/android/periodpals/ui/alert/EditAlert.kt [hunk 3]
```diff
@@ -68,45 +66,39 @@ private const val TAG = "EditAlertScreen"
...
     alertViewModel: AlertViewModel,
     navigationActions: NavigationActions,
 ) {
-  val context = LocalContext.current
-  var product by remember { mutableStateOf<Product?>(alert.product) }
-  var urgency by remember { mutableStateOf<Urgency?>(alert.urgency) }
+
+  val alert =
+      alertViewModel.selectedEditAlert.value
+          ?: return Text(text = "No alert selected", color = MaterialTheme.colorScheme.error)
```
[the remark this answers is not in the document]
Comment (francelu) [reply app/src/main/java/com/android/periodpals/ui/alert/EditAlert.kt:84]: Good thinking ! I did the changes.

File: app/src/test/java/com/android/periodpals/ui/alert/EditAlertScreenTest.kt (hunk 1 above)
Comment (francelu) [reply app/src/test/java/com/android/periodpals/ui/alert/EditAlertScreenTest.kt:73]: Done :)

File: app/src/main/java/com/android/periodpals/ui/components/AlertComponents.kt [hunk 4]
```diff
@@ -380,11 +381,40 @@ fun extractProductObject(product: Product): PeriodPalsIcon =
...
     }
+
+/**
+ * Converts a text representation of a product to a `Product` enum.
+ *
+ * @param product The text representation of the product.
+ * @return The corresponding `Product` enum, or `Product.NO_PREFERENCE` if the text does not match
+ *   any product.
+ */
+fun convertToProduct(product: String): Product =
```
[the remark this answers is not in the document]
Comment (francelu) [reply app/src/main/java/com/android/periodpals/ui/components/AlertComponents.kt:400]: Ooooh my bad, I forgot those functions were in `Alert.kt`.  
I did the changes.

File: app/src/main/java/com/android/periodpals/ui/alert/EditAlert.kt [hunk 5]
```diff
@@ -68,31 +68,28 @@ private const val TAG = "EditAlertScreen"
...
-            createdAt = ""),
     locationViewModel: LocationViewModel,
     gpsService: GPSServiceImpl,
     alertViewModel: AlertViewModel,
     navigationActions: NavigationActions,
 ) {
   val context = LocalContext.current
+
+  val alert = alertViewModel.selectedAlert.value
+  if (alert == null) {
```
[the remark this answers is not in the document]
Comment (francelu) [reply app/src/main/java/com/android/periodpals/ui/alert/EditAlert.kt:86]: Okay I'll do that, thanks !

File: app/src/main/java/com/android/periodpals/model/alert/Alert.kt [hunk 6]
```diff
@@ -51,6 +51,23 @@ enum class Status {
   SOLVED, // The alert has been resolved, help was provided
 }
 
+/** Data class representing an [icon] and [textId] for the product and urgency level */
+data class PeriodPalsIcon(val icon: Int, val textId: String)
```
Comment (lazarinibruno) [inline app/src/main/java/com/android/periodpals/model/alert/Alert.kt:55]: **Question**: wasn't this already done in a previous PR?

File: app/src/main/java/com/android/periodpals/model/alert/Alert.kt [hunk 7]
```diff
@@ -74,25 +91,35 @@ fun stringToProduct(productText: String): Product? {
...
-val LIST_OF_URGENCIES =
-    listOf(
-        PeriodPalsIcon(R.drawable.urgency_1, "Low"),
-        PeriodPalsIcon(R.drawable.urgency_2, "Medium"),
-        PeriodPalsIcon(R.drawable.urgency_3, "High"),
-    )
+/**
+ * Extracts the PeriodPalsIcon from the Alert urgency enum.
+ *
+ * @param urgency The (enum) urgency associated with the alert.
```
Comment (lazarinibruno) [inline app/src/main/java/com/android/periodpals/model/alert/Alert.kt:117]: **Nitpick**: Not necessary to specify that it is an enum

File: app/src/main/java/com/android/periodpals/ui/alert/EditAlert.kt [hunk 8]
```diff
@@ -154,6 +151,7 @@ fun EditAlertScreen(
       // Message field
       MessageField(text = message, onValueChange = { message = it })
 
+      // Start button, or reset and stop buttons
```
Comment (lazarinibruno) [inline app/src/main/java/com/android/periodpals/ui/alert/EditAlert.kt:154]: **Nitpick** : I think you did too much timer VM @francelu hahaha. This should be "Save, delete or resolve button".

**Review by lazarinibruno — APPROVED** [review]
# Summary

Thank you for this PR! It is pretty cool to be finally able to edit the alerts 😼.

## Code quality

- The code is readable and well-written.
  
- Very clever way to handle selecting the alert!
  

## Documentation

- Thanks for completing the missing documentation!
  

## Testing

- You wrote tests for the new features, so no complaints here.
  
## Steps before approval

Very solid work @francelu. LGTM.

File: app/src/main/java/com/android/periodpals/ui/alert/EditAlert.kt (hunk 8 above)
Comment (francelu) [reply app/src/main/java/com/android/periodpals/ui/alert/EditAlert.kt:154]: Hahahahhaha truuuue
I mixed it with timer VM my baaaad 😂

File: app/src/main/java/com/android/periodpals/model/alert/Alert.kt (hunk 7 above)
Comment (francelu) [reply app/src/main/java/com/android/periodpals/model/alert/Alert.kt:117]: Done !

**Pull request merged**